addons/stock_picking_batch/models/stock_move_line.py

```diff
@@ -89,3 +91,258 @@ def _add_to_wave(self, wave=False):
+                if picking_type.batch_group_by_src_loc:
+                    domain = expression.AND([domain, [('picking_ids.location_id', 'in', lines.location_id.ids)]])
+                if picking_type.batch_group_by_dest_loc:
+                    domain = expression.AND([domain, [('picking_ids.location_dest_id', 'in', lines.location_dest_id.ids)]])
+
+                potential_waves = self.env['stock.picking.batch'].search(domain)
+                wave_to_new_lines = defaultdict(set)
+
+                # These dictionaries are used to enforce batch max lines/transfers/weight limits
+                # Each time a line is matched to a wave, we update the corresponding values
+                wave_to_new_moves = defaultdict(set)
+                waves_to_new_pickings = defaultdict(set)
+                waves_new_extra_weight = defaultdict(float)
+
+                waves_nearest_parent_locations = defaultdict(int)
+                if picking_type.wave_group_by_location:
+                    valid_wave_ids = set()
+                    # We want to find the most descendant location in the wave locations list that is a parent of all the lines in each wave.
+                    # We also want to exclude waves that have lines that are not in these locations.
+                    for wave in potential_waves:
+                        for wave_location in reversed(picking_type.wave_location_ids):
+                            if all(loc._child_of(wave_location) for loc in wave.move_line_ids.location_id):
+                                waves_nearest_parent_locations[wave] = wave_location.id
+                                valid_wave_ids.add(wave.id)
+                                break
+                    potential_waves = self.env['stock.picking.batch'].browse(valid_wave_ids)
+
+                for line in lines:
+                    wave_found = False
+                    for wave in potential_waves:
+                        if line.company_id != wave.company_id \
+                        or (picking_type.batch_group_by_partner and line.move_id.partner_id != wave.picking_ids.partner_id) \
+                        or (picking_type.batch_group_by_destination and line.move_id.partner_id.country_id != wave.picking_ids.partner_id.country_id) \
+                        or (picking_type.batch_group_by_src_loc and line.location_id != wave.picking_ids.location_id) \
+                        or (picking_type.batch_group_by_dest_loc and line.location_dest_id != wave.picking_ids.location_dest_id) \
+                        or (picking_type.wave_group_by_product and line.product_id != wave.move_line_ids.product_id) \
+                        or (picking_type.wave_group_by_category and line.product_id.categ_id != wave.move_line_ids.product_categ_id) \
+                        or (picking_type.wave_group_by_location and waves_nearest_parent_locations[wave] != nearest_parent_locations[line].id):
+                            continue
```

Comment:
I think you could create a dict with all the attribute before the loop to avoid the double loop with a filter

Reply:
Could you give an example? I can't see how we can get rid of the nested loop